src/code/audio/audio_effects.c

```diff
@@ -1,21 +1,342 @@
+    pitchChange = AudioEffects_GetVibratoPitchChange(vib) + 32768.0f;
+    scaledExtent = vib->extent / 4096.0f;
+    extent = scaledExtent + 1.0f;
+    invExtent = 1.0f / extent;
+
+    // inverse linear interpolation
+    result = 1.0f / ((extent - invExtent) * pitchChange / 65536.0f + invExtent);
+
+    activeVibratoFreqScaleSum += result;
+    activeVibratoCount++;
+
+    return result;
+}
+
+void AudioEffects_NoteVibratoUpdate(Note* note) {
+    if (note->playbackState.portamento.mode != 0) {
+        note->playbackState.portamentoFreqScale = AudioEffects_GetPortamentoFreqScale(&note->playbackState.portamento);
+    }
+    if (note->playbackState.vibratoState.active) {
+        note->playbackState.vibratoFreqScale = AudioEffects_GetVibratoFreqScale(&note->playbackState.vibratoState);
+    }
+}
+
+void AudioEffects_NoteVibratoInit(Note* note) {
+    VibratoState* vib;
+    VibratoSubStruct* vibSubStruct;
+    NotePlaybackState* playbackState;
+
+    playbackState = &note->playbackState;
+
+    vib = &note->playbackState.vibratoState;
+
+    vib->active = true;
+
+    vib->curve = gWaveSamples[2];
+    if (playbackState->parentLayer->unk_0A.s.bit_3 == 1) {
+        vib->vibSubStruct = &playbackState->parentLayer->channel->vibrato;
+    } else {
+        vib->vibSubStruct = &playbackState->parentLayer->vibrato;
+    }
+
+    vibSubStruct = vib->vibSubStruct;
+    if ((vib->extentChangeTimer = vibSubStruct->vibratoExtentChangeDelay) == 0) {
+        vib->extent = (s32)vibSubStruct->vibratoExtentTarget;
+    } else {
+        vib->extent = (s32)vibSubStruct->vibratoExtentStart;
+    }
+
+    if ((vib->rateChangeTimer = vibSubStruct->vibratoRateChangeDelay) == 0) {
+        vib->rate = (s32)vibSubStruct->vibratoRateTarget;
+    } else {
+        vib->rate = (s32)vibSubStruct->vibratoRateStart;
+    }
+    playbackState->vibratoFreqScale = 1.0f;
+    vib->time = 0;
+    vib->delay = vibSubStruct->vibratoDelay;
+}
+
+void AudioEffects_NotePortamentoInit(Note* note) {
+    note->playbackState.portamentoFreqScale = 1.0f;
+    note->playbackState.portamento = note->playbackState.parentLayer->portamento;
+}
+
+void AudioEffects_AdsrInit(AdsrState* adsr, AdsrEnvelope* envelope, s16* volOut) {
+    adsr->action.asByte = 0;
+    adsr->delay = 0;
+    adsr->envelope = envelope;
+    adsr->sustain = 0.0f;
+    adsr->current = 0.0f;
+    adsr->velocity = 0.0f;
+    // (An older versions of the audio engine used in Super Mario 64 did
+    // adsr->volOut = volOut. That line and associated struct member were
+    // removed, but the function parameter was forgotten and remains.)
+}
+
+f32 AudioEffects_AdsrUpdate(AdsrState* adsr) {
+    u8 state = adsr->action.s.state;
+
+    switch (state) {
+        case ADSR_STATE_DISABLED:
+            return 0.0f;
+
+        case ADSR_STATE_INITIAL: {
+            if (adsr->action.s.hang) {
+                adsr->action.s.state = ADSR_STATE_HANG;
+                break;
+            }
+            // fallthrough
+        }
```

Comment:
Are the braces necessary here?
```suggestion
        case ADSR_STATE_INITIAL:
            if (adsr->action.s.hang) {
                adsr->action.s.state = ADSR_STATE_HANG;
                break;
            }
            // fallthrough
```